fix(studio): more resilient saving to file

panzarino opened the pull request:
<!-- Thanks for contributing! PLEASE...
- Read our contributing guidelines: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
- Read our Code Review Checklist on coding standards and what needs to be done before a PR can be merged: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#Code-Review-Checklist
- Mark this PR as "Draft" if it is not ready for review.
- Make sure you set the correct base branch based on what packages you're changing: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#branches
-->

- Closes #14767, GROW-227

### User facing changelog
Reduced occurrence of "Studio failed to save commands" error message

### Additional details
Sometimes studio would fail to save tests when there was no source maps or way to get the invocation details for a test/suite. These changes introduce a backup saving algorithm that matches based on the title of the test/suite in the case that the test cannot be found from the invocation details.

The backup plan will search through all of the test files in the spec to match by test/suite title. This still isn't a perfect solution since there could be two tests with the same name, in which case it doesn't save anything right now (as opposed to accidentally saving to the wrong test). A potential improvement to this is to match based on the `_testPath` (which includes the name of all parent suites), but that also can be problematic since it won't be able to find tests that aren't directly nested in the AST (ex. if there's a function that calls the `it` from a different file) and there's no `_testPath` for suites. We could also attempt to match by test body in addition to title but that seems more trouble than its worth right now (plus suites don't have a `body`). Perhaps we can add that in another update before Studio goes GA. Combined with copy to clipboard as a backup, this should hopefully make the Studio experience much better for users getting that error message.

I also cleaned up how Studio state gets persisted mid-test (eg. when a user enters a different url and the browser needs to reload) which turned out to not be relevant to this update but nice to have (which is why there's a bunch of updates to unrelated tests)

### How has the user experience changed?
Hopefully less error messages

### PR Tasks
<!-- These tasks must be completed before a PR is merged.
Delete tasks if they are not applicable. -->

- [x] Have tests been added/updated?
- [x] Has the original issue or this PR been tagged with a release in ZenHub? <!-- (internal team only)-->

panzarino committed:
fix(studio): more resilient saving

panzarino committed:
Merge branch 'develop' into GROW-227

panzarino committed:
Add tests for writing to runnable by title

panzarino committed:
Fix studio runner tests

panzarino committed:
Thanks stop-only!

panzarino committed:
Add studio tests without source maps

panzarino committed:
Update studio runner tests

panzarino committed:
Symlinks don't work in server tests :(

panzarino committed:
Merge branch 'develop' into GROW-227

panzarino committed:
Update save error message

panzarino committed:
Fix spec writer tests for update

panzarino committed:
Merge branch 'develop' into GROW-227

chrisbreiding commented on packages/runner-shared/src/event-manager.js:
```diff
@@ -482,13 +484,23 @@ export const eventManager = {
+    Cypress.on('test:before:run:async', (attr, test) => {
       if (studioRecorder.suiteId) {
         studioRecorder.setTestId(test.id)
       }
 
-      if (studioRecorder.hasRunnableId && test.invocationDetails) {
-        studioRecorder.setFileDetails(test.invocationDetails)
+      if (studioRecorder.hasRunnableId) {
+        if (test.invocationDetails) {
+          studioRecorder.setFileDetails(test.invocationDetails)
+        }
+
+        if (studioRecorder.suiteId) {
+          if (test.parent && test.parent.id !== 'r1') {
+            studioRecorder.setRunnableTitle(test.parent.title)
+          }
+        } else {
+          studioRecorder.setRunnableTitle(test.title)
+        }
       }
```
I think it would be good to abstract these lines into a `studioRecorder` method or another utility. We should try to keep the responsibility of this file to managing events and move any logic into separate places.

chrisbreiding replied:
Could you move these lines (486-502) into `studioRecorder` as well? 

I think the litmus test for abstracting out of this file is whether there is `if/else` logic involved. There are a couple other places in this file that still have `if/else`, but they're unrelated to this PR, so of course don't need to be taken care of now.

chrisbreiding commented on packages/runner-shared/src/event-manager.js:
```diff
@@ -555,35 +567,30 @@ export const eventManager = {
-    if (state.studioSuiteId) {
-      studioRecorder.setSuiteId(state.studioSuiteId)
-    }
+    if (studio) {
+      if (studio.testId) {
+        studioRecorder.setTestId(studio.testId)
+      }
+
+      if (studio.suiteId) {
+        studioRecorder.setSuiteId(studio.suiteId)
+      }
 
-    if (state.studioUrl) {
-      studioRecorder.setUrl(state.studioUrl)
+      if (studio.url) {
+        studioRecorder.setUrl(studio.url)
+      }
```
Another instance we could abstract into a `studioRecorder` method.

panzarino committed:
Move init studio methods into recorder

panzarino committed:
Combine studio init fns

panzarino commented:
@chrisbreiding updated

panzarino committed:
Merge branch 'develop' into GROW-227

panzarino committed:
Pull more methods into studio recorder and add tests

panzarino commented:
@chrisbreiding Oops, missed that one - I moved that over now and added some extra unit tests for good measure (in addition to being tested through integration and e2e tests)

chrisbreiding reviewed (approved):
Looks good 👍

panzarino committed:
Merge branch 'develop' into GROW-227

panzarino committed:
Merge branch 'develop' into GROW-227
